confgenerator/confmerger.go

```diff
@@ -0,0 +1,230 @@
+      type: google_cloud_logging
+  service:
+    pipelines:
+      default_pipeline:
+        receivers: [syslog]
+        exporters: [google]
+
+metrics:
+  receivers:
+    hostmetrics:
+      type: hostmetrics
+      collection_interval: 60s
+  exporters:
+    google:
+      type: google_cloud_monitoring
+  service:
+    pipelines:
+      default_pipeline:
+        receivers: [hostmetrics]
+        exporters: [google]`,
+
+		"windows": `logging:
+  receivers:
+    windows_event_log:
+      type: windows_event_log
+      channels: [System,Application,Security]
+  exporters:
+    google:
+      type: google_cloud_logging
+  service:
+    pipelines:
+      default_pipeline:
+        receivers: [windows_event_log]
+        exporters: [google]
+
+metrics:
+  receivers:
+    hostmetrics:
+      type: hostmetrics
+      collection_interval: 60s
+    mssql:
+      type: mssql
+      collection_interval: 60s
+    iis:
+      type: iis
+      collection_interval: 60s
+  exporters:
+    google:
+      type: google_cloud_monitoring
+  service:
+    pipelines:
+      default_pipeline:
+        receivers: [hostmetrics,mssql,iis]
+        exporters: [google]`,
+	}
+)
+
+func MergeConfFiles(builtInConfPath, userConfPath, mergedConfPath, platform string) error {
+
+	// Write the built-in conf to disk for debugging purpose.
+	if err := os.MkdirAll(path.Dir(builtInConfPath), 0777); err != nil {
+		return fmt.Errorf("failed to create the folder for the built-in config file: %w \n", err)
+	}
+	f, err := os.Create(builtInConfPath)
+	if err != nil {
+		return fmt.Errorf("failed to create the built-in config file %q: %w \n", builtInConfPath, err)
+	}
+	defer f.Close()
+	if _, err = f.WriteString(builtInConf[platform]); err != nil {
+		return fmt.Errorf("failed to write to the built-in config file %q: %w \n", builtInConfPath, err)
+	}
+
+	// Read the built-in config file.
+	original, err := ReadUnifiedConfigFromFile(builtInConfPath)
+	if err != nil {
+		return fmt.Errorf("failed to read config from the built-in config file %q: %w \n", builtInConfPath, err)
+	}
+
+	// Optionally merge the user config file.
+	if _, err = os.Stat(userConfPath); err != nil {
+		if os.IsNotExist(err) {
+			// Skip the merge if the user config file does not exist.
+		} else {
+			return fmt.Errorf("failed to retrieve the user config file %q: %w \n", userConfPath, err)
+		}
+	} else {
+		overrides, err := ReadUnifiedConfigFromFile(userConfPath)
+		if err != nil {
+			return err
+		}
+		mergeConfigs(&original, &overrides)
+	}
+
+	// Write the merged conf file.
+	configBytes, err := yaml.Marshal(original)
+	if err != nil {
+		return fmt.Errorf("failed to convert the merged config %q to yaml: %w \n", mergedConfPath, err)
+	}
+	if err := ioutil.WriteFile(mergedConfPath, configBytes, 0644); err != nil {
+		return fmt.Errorf("failed to write the merged config file %q: %w \n", mergedConfPath, err)
+	}
+	return nil
+}
+
+// TODO(lingshi): Refactor this part to de-duplicate repeated code.
```

Comment:
Similar here, I'm working on the side to refactor this part while this PR is being reviewed. But it's not a GA blocker, so if the PR is ready to ship before I get there, I will send a followup PR.

Reply:
Done

Reply:
Done